rectools/metrics/debias.py

```diff
@@ -0,0 +1,166 @@
+                "and specify `is_debiased` as `True` "
+                "or otherwise use `calc` and `calc_per_user` methods for auto de-biasing."
+            )
+
+    def make_debias(self, interactions: pd.DataFrame) -> pd.DataFrame:
+        """
+        Downsample the size of interactions, excluding some interactions with popular items.
+
+        Algorithm:
+
+            1. Calculate item "popularity"
+            (here: number of unique users that had interaction with the item) distribution from interactions;
+            2. Find first (Q1) and third (Q3) quartiles in items "popularity" distribution;
+            3. Calculate IQR = Q3 - Q1;
+            4. Calculate maximum value inside by formula: Q3 + iqr_coef * IQR;
+            5. Down-sample for all exceeding items in interactions,
+            randomly keeping the maximum group of users to a size not exceeding
+            maximum value inside
+
+        Parameters
+        ----------
+        interactions : pd.DataFrame
+            Table with previous user-item interactions,
+            with columns `Columns.User`, `Columns.Item`.
+        iqr_coef : float, default 1.5
+            Coefficient for defining as the maximum value inside the border.
+        random_state: float, default 32
+            Pseudorandom number generator state to control the down-sampling.
+
+        Returns
+        -------
+        pd.DataFrame
+            Downsampling interactions.
+        """
+        if len(interactions) == 0:
+            return interactions
+
+        interactions_for_debiasing = interactions.copy()
+
+        num_users_interacted_with_item = interactions_for_debiasing.drop_duplicates(subset=Columns.UserItem)[
+            Columns.Item
+        ].value_counts()
+
+        quantiles = num_users_interacted_with_item.quantile(q=[0.25, 0.75])
+        q1, q3 = quantiles.loc[0.25], quantiles.loc[0.75]
+        iqr = q3 - q1
+        max_border = int(q3 + self.debias_config.iqr_coef * iqr)
+
+        item_outside_max_border = num_users_interacted_with_item[num_users_interacted_with_item > max_border].index
+
+        mask_outside_max_border = interactions_for_debiasing[Columns.Item].isin(item_outside_max_border)
+        interactions_result = interactions_for_debiasing[~mask_outside_max_border]
+        interactions_downsampling = interactions_for_debiasing[mask_outside_max_border]
+
+        interactions_downsampling = (
+            interactions_downsampling.sample(frac=1.0, random_state=self.debias_config.random_state)
+            .groupby(Columns.Item)
+            .head(max_border)
+        )
+
+        result_dfs = [interactions_result, interactions_downsampling]
+        interactions_result = pd.concat(result_dfs, ignore_index=True)
+
+        return interactions_result
+
+
+def calc_debias_for_fit_metrics(
+    metrics: tp.Dict[str, DebiasableMetrikAtK], interactions: pd.DataFrame
+) -> tp.Dict[DebiasConfig, tp.List[tp.Union[int, pd.DataFrame]]]:
```

Comment:
If you need to put exactly 2 objects, especially with different types, you should definitely use tuple, never list
```suggestion
) -> tp.Dict[DebiasConfig, tp.Tuple[int, pd.DataFrame]]:
```

Reply:
I used a sheet because the values within it could change as you go through the metrics.
I added functionality: transforming tuple into list -> changing values in it -> transforming list into tuple.

Reply:
Let's talk about the whole method again. How about:
1. return `Tuple[DebiasConfig, int, pd.DataFrame]` where int is for max_k and pd.DataFrame is for debiased interactions. we don't need a dict right now so we can simplify it here
2. Let's call this `calc_debiased_fit_task`. This is used only for metrics with `fit` method because we need max_k only for `fit` and nowhere else.
3. Let's do smth like:
```
configs = set([metric.debias_config for metric in metrics])
conf_to_k = {config: [] for config in configs}
for metric in metrics:
    conf_to_k[metric.config].append(metric.k)
max_k_for_config = {config: max(k_requests) for config, k_requests in conf_to_k.items()}
res = ((config, max_k_for_config[config], DebiasableMetricAtK.make_debias(interactions, config) for config in configs)
return res
```
Here we make `make_debias` a classmethod which accepts debias config as an argument. And it is actually quite useful for external users. We can even make it a separate method which is even more useful. I would just name it `debias_interactions`

Reply:
> we don't need a dict right now so we can simplify it here

Having a dict is good to show that we don't have 2 items with the same configs

> `set([x for x in xx])` 

`set(x for x in xx)` 

> `k_requests`

Sounds like we are counting requests. Let's keep just `k` or `requested_k`

Agreed with the rest